### PR TITLE
More custom geometry

### PR DESCRIPTION
**Pull request by codexhound**
-Added shape creation functions in custom geometry
-Added shape protrusion
-Added circle generation
-Added circle graph generation
-Added sphere generation
-Added square generation
-Added mesh generation from vector of points

Math
-Added vector rotation
-Added degrees to radians and radians to degrees
-Added average function

### CODE REVIEW

File: Source/Urho3D/Graphics/CustomGeometry.h [hunk 1]
```diff
@@ -66,6 +67,34 @@ class URHO3D_API CustomGeometry : public Drawable
     /// Draw to occlusion buffer. Return true if did not run out of triangles.
     bool DrawOcclusion(OcclusionBuffer* buffer) override;
 
+    /// Geometry Shape Helper Functions (normal defaults to (0,1,0)
+
+    /// Shape Creation Functions, can be used as input into protrude shape
+    static std::vector<Vector3> GetCircleShape(float radius = 1, size_t iterations = 100, float startTheta = 0, float endTheta = 2 * M_PI);
```
Comment (ArnisLielturks) [inline Source/Urho3D/Graphics/CustomGeometry.h]: Any specific reason why not use Urho's built in containers?

**Comment by codexhound** [comment]
Not really except that is just my standard practice when coding is to use
the standard containers. I switch them out probably pretty easily.

File: Source/Urho3D/Graphics/CustomGeometry.h (hunk 1 above)
Comment (codexhound) [reply Source/Urho3D/Graphics/CustomGeometry.h]: changes uploaded

**Comment by codexhound** [comment]
Requested changes posted

File: Source/Urho3D/Graphics/CustomGeometry.cpp [hunk 2]
```diff
@@ -204,6 +204,291 @@ bool CustomGeometry::DrawOcclusion(OcclusionBuffer* buffer)
...
+        float curY1 = radius * sin(curTheta1);
+        float curX2 = radius * cos(curTheta2);
+        float curY2 = radius * sin(curTheta2);
+
+        shapeList.Push(Urho3D::Vector3(curX1, 0, curY1));
+        shapeList.Push(Urho3D::Vector3(curX2, 0, curY2));
+
+        if (i >= iterations - 1) {
+            float curTheta =0;
+            if (std::abs(endTheta - startTheta) < (2*M_PI)) {
```
Comment (ArnisLielturks) [inline Source/Urho3D/Graphics/CustomGeometry.cpp]: Urho3D::Abs can be used.

File: Source/Urho3D/Graphics/CustomGeometry.h [hunk 3]
```diff
@@ -66,6 +66,34 @@ class URHO3D_API CustomGeometry : public Drawable
...
+    /// Make the custom geometry into a circle, change start and stop to make a segment instead
+    /// Set clear=false and geomNum for making multiple circle segments
+    void MakeCircle(float radius = 1, size_t iterations = 300, float startTheta = 0, float endTheta = 2 * M_PI,
+        bool clear = true, int geomNum = 0);
+    /// Draws a shape generated by connecting the points in the input vector (the end point will connect to the start)
+    void MakeShape(const Vector<Urho3D::Vector3>& pointList, bool connectTail = true);
+    /// Makes this custom geometry into a square shape
+    void MakeSquare(float size);
+    /// Produces a circle graph given a vector of paired (weights and materials)
+    void MakeCircleGraph(const Vector<std::pair<float, Urho3D::SharedPtr<Urho3D::Material> > >& parts,
```
Comment (ArnisLielturks) [inline Source/Urho3D/Graphics/CustomGeometry.h]: Urho3D::Pair can be used.

**Comment by codexhound** [comment]
the rest of the standard containers have been replaced

**Comment by codexhound** [comment]
How does merging to master work? I've never done this before with git. Does an admin have to do the merge since I don't see a merge button on my end?

**Comment by ArnisLielturks** [comment]
It has to be done by one of the maintainers. I'll merge this today after few hours. Thanks for the PR!

**Comment by eugeneko** [comment]
I know I'm late for the party, but this PR violates code standard and has a lot of stylistic issues all over the code. Can anybody fix it?
I can do it on my own tho if nobody cares...

**Comment by ArnisLielturks** [comment]
@eugeneko My mistake, didn't paid attention to the that. Feel free to provide your feedback about all the code style issues.

**Comment by Modanung** [comment]
If I may do a little nitpicking; I noticed the lowercase `sin` and `cos` are used instead of `Sin` and `Cos`. Although just wrappers and a very minor detail, I believe it would make more sense to use the latter.

**Pull request merged**